### PR TITLE
[RHOSINFRA-81] - Added external LB

### PR DESCRIPTION
**Pull request by ovorobio**

### CODE REVIEW

**Comment by rhosqeauto** [comment]
Can one of the admins verify this patch?

File: playbooks/provisioner/virsh/main.yaml [hunk 1]
```diff
@@ -294,6 +294,14 @@
             ansible_ssh_host="{{ item }}"
         with_items: groups['openstack_nodes']
 
+      - name: update the ssh host name of Loadbalancer machine
+        add_host:
+           name="{{ item }}"
+           ansible_ssh_host="{{ item }}"
+        with_items: groups['loadbalancer']
+        ignore_errors: True
```
Comment (obaranov) [inline playbooks/provisioner/virsh/main.yaml]: I think it would be better here to use condition and call "add_host" only when loadbalancer is defined

File: roles/ospd/overcloud/loadbalancer/provision/files/vlan.modules [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+exec /sbin/modprobe 8021q >/dev/null 2>&1
```
Comment (obaranov) [inline roles/ospd/overcloud/loadbalancer/provision/files/vlan.modules]: Add empty new line at the end of the script.

File: roles/ospd/overcloud/network/isolation/three-nics-vlans/templates/vlan-internal_api.cfg.j2 [hunk 3]
```diff
@@ -0,0 +1,7 @@
+DEVICE={{ provisioner.topology.nodes.loadbalancer.network.undercloud.oc_interface }}.101
```
Comment (obaranov) [inline roles/ospd/overcloud/network/isolation/three-nics-vlans/templates/vlan-internal_api.cfg.j2]: Avoid hardcoded values if possible. It looks like these tag values can be taken form the installer.overcloud.network.template.content.parameter_defaults.*VlanID settings.

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 4]
```diff
@@ -72,3 +72,8 @@
       lineinfile:
           dest: "/etc/sudoers"
           line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "ospd/overcloud/loadbalancer/provision/", when: provisioner.topology.nodes.loadbalancer is defined }
```
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: please rebase your patch, the path has changed

Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: Also, please have it as it's own role "loadbalancer" and remove the tasks that are osp-d specific to a different role
so basically:
roles/loadbalancer - will install / configure everything needed on the dedicated VM to create the loadbalancer working
roles/installer/ospd/loadbalancer - will configure the loadbalancer according to the OSP-d specific needs / vars.
always keep in mind that future work might not be OSP-d related but still we will want to reuse what you wrote for the LB, so break anything that is not OSP-d related to it's own role

File: playbooks/provisioner/virsh/main.yaml [hunk 5]
```diff
@@ -294,6 +294,14 @@
...
         with_items: groups['openstack_nodes']
 
+      - name: update the ssh host name of Loadbalancer machine
+        add_host:
+           name="{{ item }}"
+           ansible_ssh_host="{{ item }}"
+        with_items: groups.loadbalancer
+        when: groups.loadbalancer is defined
+
+
```
Comment (tkammer) [inline playbooks/provisioner/virsh/main.yaml]: I don't think this is necessary with the latest code, please rebase and check

File: settings/provisioner/topology/loadbalancer.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
+name: "loadbalancer"
+cpu: "2"
+memory: "4096"
+type: "external"
```
Comment (tkammer) [inline settings/provisioner/topology/loadbalancer.yaml]: why is this needed here?
the definition for external LB is osp-d based no?
Haveing a dedicated VM for it already config it as such.

File: settings/provisioner/topology/loadbalancer.yaml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        dev: "/dev/vda"
+        size: "20G"
+network:
+    provision_interface: "eth0"
+    oc_interface: "eth1"
+    ex_interface: "eth2"
+    undercloud:
+        oc_interface: "eth1"
+        internal_ip: 172.17.1.248/24
+    plain:
```
Comment (tkammer) [inline settings/provisioner/topology/loadbalancer.yaml]: same, these are osp-d related variables, please move it to settings/installer/ospd/<somewhere>

**Comment by tkammer** [comment]
Overall it's a good first patch, but we need to make it more general such as we can use it with other installers (or at least most of it).
Please move the OSP-d related vars/plays to be under OSP-d and anything else that is LB specific have it in it's own role (not connected to OSP-d)

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 8]
```diff
@@ -72,3 +72,9 @@
       lineinfile:
           dest: "/etc/sudoers"
           line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "loadbalancer/provision/", ll}
```
Comment (yfried-redhat) [inline playbooks/installer/ospd/undercloud/pre.yml]: what does `ll` mean?
Also, shouldn't there be a "when" here?

File: roles/installer/ospd/loadbalancer/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+---
+- name: create VLAN interface for internal_api network
+  become_user: root
+  template: src=vlan-internal_api.cfg.j2
```
Comment (yfried-redhat) [inline roles/installer/ospd/loadbalancer/tasks/main.yml]: We prefer this syntax:

```yaml
template:
    src: vlan-internal_api.cfg.j2
    dest: "/etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}"
```

File: roles/installer/ospd/loadbalancer/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+            dest=/etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
+  when: installer.overcloud.loadbalancer.network.internal_api is defined
+
+- name: Start internal_api interface
+  shell: sudo /usr/sbin/ifup {{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
```
Comment (yfried-redhat) [inline roles/installer/ospd/loadbalancer/tasks/main.yml]: I don't know how this works for you, but it's better to quote every string that contains var reference. So:
`shell: "sudo /usr/sbin/ifup {{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}"`

File: roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+---
+- name: copy the loadbalancer template
+  template:
+    src: "external-lb.yaml.j2"
```
Comment (yfried-redhat) [inline roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml]: Please use 4 spaces for consistency.

File: roles/loadbalancer/provision/files/vlan.modules [hunk 12]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+exec /sbin/modprobe 8021q >/dev/null 2>&1
```
Comment (yfried-redhat) [inline roles/loadbalancer/provision/files/vlan.modules]: Do we really need to keep a file just for 1 line?

File: roles/loadbalancer/provision/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+#this is required to have the 22 port free on haproxy VIP's
+- name: setup ssh for loadbalancer
+  lineinfile: dest=/etc/ssh/sshd_config line="ListenAddress {{ installer.overcloud.loadbalancer.network.plain.ip | ipaddr('address') }}"
+
+- name: restart sshd service
+  service: name=sshd state=restarted
+
+- name: enable vlan kernel module
+  copy: src=vlan.modules
+        dest=/etc/sysconfig/modules/vlan.modules
```
Comment (yfried-redhat) [inline roles/loadbalancer/provision/tasks/main.yml]: `vlan.modules` is a single line. let's use it explicitly instead of storing the file in repo.

**Comment by yfried-redhat** [comment]
In playbooks/roles please use the following format

```yaml
module:
    arg: value
    arg2: "some value with {{ variable.that }} because we need {{ to.protect.jinja2.syntax }} from yaml formating"
```
Instead of this one:

```yaml
module: arg=value
              arg2=some value with {{ variable.that }} because we need {{ to.protect.jinja2.syntax }} from yaml formating
```

File: roles/installer/ospd/loadbalancer/tasks/main.yml [hunk 14]
```diff
@@ -1,10 +1,11 @@
 ---
 - name: create VLAN interface for internal_api network
   become_user: root
-  template: src=vlan-internal_api.cfg.j2
-            dest=/etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
+  template:
+  src: vlan-internal_api.cfg.j2
```
Comment (yfried-redhat) [inline roles/installer/ospd/loadbalancer/tasks/main.yml]: I think you missed indentation here.

File: playbooks/installer/ospd/overcloud/run.yml [hunk 15]
```diff
@@ -20,6 +20,7 @@
   roles:
       - {role: "installer/ospd/overcloud/storage/ceph/", when: installer.overcloud.storage.backend is defined and installer.overcloud.storage.backend == "ceph"}
       - {role: "installer/ospd/overcloud/storage/netapp/", when: installer.overcloud.storage.backend is defined and installer.overcloud.storage.backend == "netapp"}
+      - {role: "installer/ospd/loadbalancer/", when: provisioner.topology.nodes.loadbalancer is defined }
```
Comment (tkammer) [inline playbooks/installer/ospd/overcloud/run.yml]: instead of referencing a role twice, you can create a dependency between roles in the "meta" folder[1].
Please move the invocation there so we will have only one invocation rather than two.

[1] - http://docs.ansible.com/ansible/playbooks_roles.html#role-dependencies

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 16]
```diff
@@ -57,18 +57,24 @@
...
+
+      - name: add user to sudoers
+        lineinfile:
+            dest: "/etc/sudoers"
+            line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "loadbalancer/provision/", when: provisioner.topology.nodes.loadbalancer is defined }
```
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: this should only call the role "loadbalancer" which in turn, will call the provision. (through meta)

File: roles/installer/ospd/loadbalancer/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,11 @@
+---
+- name: create VLAN interface for internal_api network
+  become_user: root
+  template:
+  src: vlan-internal_api.cfg.j2
```
Comment (tkammer) [inline roles/installer/ospd/loadbalancer/tasks/main.yml]: this will fail as it's not indented. please indent 4 spaces

File: roles/installer/ospd/loadbalancer/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
```
Comment (tkammer) [inline roles/installer/ospd/loadbalancer/tasks/main.yml]: same

File: roles/installer/ospd/loadbalancer/templates/vlan-internal_api.cfg.j2 [hunk 19]
```diff
@@ -0,0 +1,7 @@
+DEVICE={{ installer.overcloud.loadbalancer.network.undercloud.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
+BOOTPROTO=static
+ONBOOT=yes
+IPADDR={{ installer.overcloud.loadbalancer.network.undercloud.internal_ip | ipaddr('address') }}
+PREFIX={{ installer.overcloud.loadbalancer.network.undercloud.internal_ip  | ipaddr('prefix') }}
+VLAN=yes
+NM_CONTROLLED=no
```
Comment (tkammer) [inline roles/installer/ospd/loadbalancer/templates/vlan-internal_api.cfg.j2]: new line at the end

File: roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml [hunk 20]
```diff
@@ -0,0 +1,16 @@
+---
+- name: copy the loadbalancer template
+  template:
+      src: "external-lb.yaml.j2"
+      dest: "~/external-lb.yaml"
+
+- name: append the External LB template line to the base overcloud deploy script
+  lineinfile:
+      dest: "~/overcloud_deploy.sh"
+      line: '                       -e /usr/share/openstack-tripleo-heat-templates/environments/external-loadbalancer-vip.yaml \\'
```
Comment (tkammer) [inline roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml]: ansible 2 has removed the need for "\\", this will actually fail in ansible 2, please remove one of the "\".

File: roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+
+
+- name: append External LB config to the base overcloud deply script
+  lineinfile:
+      dest: "~/overcloud_deploy.sh"
+      line: '                       -e ~/external-lb.yaml\\'
```
Comment (tkammer) [inline roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,92 @@
+- name: switch selinux to permissive
+  selinux: policy=targeted state=permissive
+
+#this is required to have the 22 port free on haproxy VIP's
+- name: setup ssh for loadbalancer
+  lineinfile: dest=/etc/ssh/sshd_config line="ListenAddress {{ installer.overcloud.loadbalancer.network.plain.ip | ipaddr('address') }}"
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: please use the format:
lineinfile: 
    dest: <> 
    line: <> 
etc...

File: roles/loadbalancer/provision/tasks/main.yml [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+
+- name: restart sshd service
+  service: name=sshd state=restarted
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+
+- name: enable vlan kernel module
+  linefile:
+  create: yes
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: please add indentation of 4 spaces

File: roles/loadbalancer/provision/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+  state: present
+  line: "exec /sbin/modprobe 8021q >/dev/null 2>&1"
+  mode: "0700"
+  owner: "root"
+  dest: /etc/sysconfig/modules/vlan.modules
+
+- name: create VLAN interface for plain network
+  template:
+  src: vlan-plain.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.oc_interface }}
+  when: installer.overcloud.loadbalancer.network.plain is defined
+
+- name: create VLAN interface for external network
+  template:
+  src: vlan-external.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.ex_interface }}
+  when: installer.overcloud.loadbalancer.network.external is defined
+
+- name: create VLAN interface for internal_api network
+  template:
+  src: vlan-internal_api.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.InternalApiNetworkVlanID }}
+  when: installer.overcloud.loadbalancer.network.internal_api is defined
+
+- name: create VLAN interface for storage network
+  template:
+  src: vlan-storage.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.StorageNetworkVlanID }}
+  when: installer.overcloud.loadbalancer.network.storage is defined
+
+- name: create VLAN interface for storage_mgmnt network
+  template:
+  src: vlan-storage_mgmt.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+  dest: /etc/sysconfig/network-scripts/ifcfg-{{ installer.overcloud.loadbalancer.network.oc_interface }}.{{ installer.overcloud.network.template.content.parameter_defaults.StorageMgmtNetworkVlanID }}
+  when: installer.overcloud.loadbalancer.network.storage_mgmnt is defined
+
+- name: create VLAN interface for tenant network
+  template:
+  src: vlan-tenant.cfg.j2
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+
+- name: Install release tool on loadbancer
+  command: "yum localinstall -y {{ product.rpm }}"
+
+- name: Execute rhos-release for OSP-Director
+  command: rhos-release {{ installer.version.major}}
+
+- name: update all packages
+  yum:
+  name: *
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: same

File: roles/loadbalancer/provision/tasks/main.yml [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+  state: latest
+
+- name: install haproxy service
+  yum:
+  name: haproxy
```
Comment (tkammer) [inline roles/loadbalancer/provision/tasks/main.yml]: well, I can keep writing "same" but you get the point :)
please go over all your playbooks and make sure they are indented :)

File: settings/installer/ospd/loadbalancer/external.yml [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+                interfaces:
+                    management:
+                        label: "eth0"
+                    data:
+                        label: "eth1"
+                    external:
+                        label: "eth2"
+            controllers:
+                pool_start: 251
+                name_prefix: "overcloud-controller-"
```
Comment (tkammer) [inline settings/installer/ospd/loadbalancer/external.yml]: new line

**Comment by rhosqeauto** [comment]
Can one of the admins verify this patch?

**Comment by rhosqeauto** [comment]
Can one of the admins verify this patch?

**Comment by rhosqeauto** [comment]
Can one of the admins verify this patch?

**Comment by rhosqeauto** [comment]
Can one of the admins verify this patch?

File: playbooks/installer/ospd/overcloud/run.yml (hunk 15 above)
Comment (tkammer) [inline playbooks/installer/ospd/overcloud/run.yml]: please change: when: provisioner.topology.nodes.loadbalancer is defined
to: when: "'loadbalancer' in groups"

File: playbooks/installer/ospd/undercloud/pre.yml (hunk 16 above)
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: Still forgot this comment....

**Comment by yfried-redhat** [comment]
retest

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 34]
```diff
@@ -72,3 +72,9 @@
       lineinfile:
           dest: "/etc/sudoers"
           line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "loadbalancer/provision/", when: 'loadbalancer' in groups}
```
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: please change to "loadbalancer" (remove the "provision")

**Comment by yfried-redhat** [comment]
I think this needs a rebase

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 35]
```diff
@@ -72,3 +72,9 @@
       lineinfile:
           dest: "/etc/sudoers"
           line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "loadbalancer/provision/", when: groups.loadbalancer is defined }
```
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: remove "provision", leave it as:
`- { role: "loadbalancer", when: groups.loadbalancer is defined }`

File: roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml [hunk 36]
```diff
@@ -0,0 +1,16 @@
+---
```
Comment (tkammer) [inline roles/installer/ospd/overcloud/network/isolation/three-nics-vlans/tasks/loadbalancer.yml]: this file should be taken out and put under "installer/ospd/loadbalancer"

File: playbooks/installer/ospd/undercloud/pre.yml [hunk 37]
```diff
@@ -72,3 +72,9 @@
       lineinfile:
           dest: "/etc/sudoers"
           line: "{{ installer.user.name }} ALL=(root) NOPASSWD:ALL"
+
+- name: Deploy and configure external loadbalancer instance
+  hosts: loadbalancer
+  roles:
+    - { role: "loadbalancer/", when: groups.loadbalancer is defined }
```
Comment (tkammer) [inline playbooks/installer/ospd/undercloud/pre.yml]: remove "/" from role name

**Comment by yfried-redhat** [comment]
retest
